In [1]:
import numpy as np
import pandas as pd

# 一、索引器

### 1. 表的列、行索引 
> Series = dataframe\[列名\] == dataframe.无空格列名
>
> dataframe = dataframe\[列名列表\]
>
> ------------
>
> 以字符串为索引：
>
> 值 / Series = dataframe\[item\] (多个索引 --- item -> items列表)
>
> 索引集合中唯一出现的索引，可以使用切片，两个顶点都包含
>
> 以整数为索引的 Series: \[int\], \[int_list\]


In [27]:
# > Series = dataframe\[列名\] == dataframe.无空格列名
df = pd.read_csv('data/learn_pandas.csv', usecols = ['School', 'Grade', 'Name', 'Gender', 'Weight', 'Transfer'])
df['Name'].head(6)

0      Gaopeng Yang
1    Changqiang You
2           Mei Sun
3      Xiaojuan Sun
4       Gaojuan You
5       Xiaoli Qian
Name: Name, dtype: object

In [21]:
df.Name.head(6)

0      Gaopeng Yang
1    Changqiang You
2           Mei Sun
3      Xiaojuan Sun
4       Gaojuan You
5       Xiaoli Qian
Name: Name, dtype: object

In [28]:
# > dataframe = dataframe\[列名列表\]
df[['Gender', 'Name']].head(6)

Gender            Name
0  Female    Gaopeng Yang
1    Male  Changqiang You
2    Male         Mei Sun
3  Female    Xiaojuan Sun
4    Male     Gaojuan You
5  Female     Xiaoli Qian

In [36]:
# > 以字符串为索引：
# >
# > 值 / Series = dataframe\[item\] (多个索引 --- item -> items列表)
df_ = pd.Series([1,2,3,4,5,6], index = ['a', 'b', 'a', 'a', 'a', 'c'])
df_['a'], df_['b'], df_[['c', 'b']], df_['c': 'b': -2]   # > 索引集合中唯一出现的索引，可以使用切片，两个顶点都包含

(a    1
 a    3
 a    4
 a    5
 dtype: int64,
 2,
 c    6
 b    2
 dtype: int64,
 c    6
 a    4
 b    2
 dtype: int64)

In [35]:
# > 以整数为索引的 Series: [int], [int_list]
df_[1], df_[[2, 3]], df_[1:-1:2]

(2,
 a    3
 a    4
 dtype: int64,
 b    2
 a    4
 dtype: int64)

### 2. loc 索引器
> iloc: 基于位置
>
> loc: 基于元素: loc\[*\] -- \*:
>
> loc\[行\]  ， loc\[行， 列\]
>
> set_index: 设置索引
>
> 单元素
>
> 元素列表
>
> 元素切片
>
> 布尔列表:
> 
> --- 1. operator >
>
> --- 2. isin
>
> --- 3. |, &, ~
>
> 函数
>





In [37]:
df_demo = df.set_index('Name')
df_demo.head()



School      Grade  Gender  Weight  \
Name                                                                       
Gaopeng Yang    Shanghai Jiao Tong University   Freshman  Female    46.0   
Changqiang You              Peking University   Freshman    Male    70.0   
Mei Sun         Shanghai Jiao Tong University     Senior    Male    89.0   
Xiaojuan Sun                 Fudan University  Sophomore  Female    41.0   
Gaojuan You                  Fudan University  Sophomore    Male    74.0   

               Transfer  
Name                     
Gaopeng Yang          N  
Changqiang You        N  
Mei Sun               N  
Xiaojuan Sun          N  
Gaojuan You           N

In [41]:
# * 为单个元素
print(df_demo.loc['Qiang Sun'])  # dataframe
print(df_demo.loc['Quan Zhao'])  # series

# * 为单个元素，*
print(df_demo.loc['Qiang Sun', 'School'])

                                  School      Grade  Gender  Weight Transfer
Name                                                                        
Qiang Sun            Tsinghua University     Junior  Female    53.0        N
Qiang Sun            Tsinghua University  Sophomore  Female    40.0        N
Qiang Sun  Shanghai Jiao Tong University     Junior  Female     NaN        N
School      Shanghai Jiao Tong University
Grade                              Junior
Gender                             Female
Weight                                 53
Transfer                                N
Name: Quan Zhao, dtype: object
Name
Qiang Sun              Tsinghua University
Qiang Sun              Tsinghua University
Qiang Sun    Shanghai Jiao Tong University
Name: School, dtype: object


In [43]:
# * 为元素列表， *
df_demo.loc[['Qiang Sun', 'Quan Zhao'], ['School', 'Gender']]

School  Gender
Name                                            
Qiang Sun            Tsinghua University  Female
Qiang Sun            Tsinghua University  Female
Qiang Sun  Shanghai Jiao Tong University  Female
Quan Zhao  Shanghai Jiao Tong University  Female

In [49]:
# * 切片， *
print(df_demo.loc['Gaojuan You':'Gaoqiang Qian', 'School':'Gender'])

# 对整数索引， 包含端点、起点、不允许重复值
df_loc_slice_demo = df_demo.copy()
df_loc_slice_demo.index = range(df_demo.shape[0], 0, -1)
print(list(range(df_demo.shape[0], 0, -1))[:20])
print(df_loc_slice_demo.loc[5: 3])
print(df_loc_slice_demo.loc[3: 5])  # [3, 5] 不是整数位置切片

                                      School      Grade  Gender
Name                                                           
Gaojuan You                 Fudan University  Sophomore    Male
Xiaoli Qian              Tsinghua University   Freshman  Female
Qiang Chu      Shanghai Jiao Tong University   Freshman  Female
Gaoqiang Qian            Tsinghua University     Junior  Female
[200, 199, 198, 197, 196, 195, 194, 193, 192, 191, 190, 189, 188, 187, 186, 185, 184, 183, 182, 181]
                          School   Grade  Gender  Weight Transfer
5               Fudan University  Junior  Female    46.0        N
4            Tsinghua University  Senior  Female    50.0        N
3  Shanghai Jiao Tong University  Senior  Female    45.0        N
Empty DataFrame
Columns: [School, Grade, Gender, Weight, Transfer]
Index: []


In [50]:
# > 布尔列表:
# > 
# > --- 1. operator >
# >
# > --- 2. isin
# >
# > --- 3. |, &, ~
# * 布尔列表, 选中 True 的 位置
print(df_demo.loc[df_demo.Weight > 70].head())
# series.isin(list)
print(df_demo.loc[df.demo.Grade.isin(['Freshman', 'Senior'])])


School      Grade Gender  Weight  \
Name                                                                     
Mei Sun        Shanghai Jiao Tong University     Senior   Male    89.0   
Gaojuan You                 Fudan University  Sophomore   Male    74.0   
Xiaopeng Zhou  Shanghai Jiao Tong University   Freshman   Male    74.0   
Xiaofeng Sun             Tsinghua University     Senior   Male    71.0   
Qiang Zheng    Shanghai Jiao Tong University     Senior   Male    87.0   

              Transfer  
Name                    
Mei Sun              N  
Gaojuan You          N  
Xiaopeng Zhou        N  
Xiaofeng Sun         N  
Qiang Zheng          N

In [53]:
# 复旦大学中体重超过70kg的大四学生，或者北大男生中体重超过80kg的非大四的学生：
condition_1_1 = df_demo.School == 'Fudan University'
condition_1_2 = df_demo.Grade == 'Senior'
condition_1_3 = df_demo.Weight > 70
condition_1 = condition_1_1 & condition_1_2 & condition_1_3
condition_2_1 = df_demo.School == 'Peking University'
condition_2_2 = df_demo.Grade == 'Senior'
condition_2_3 = df_demo.Weight > 80
condition_2 = condition_2_1 & (~condition_2_2) & (condition_2_3)
df_demo.loc[condition_1 | condition_2]


School     Grade Gender  Weight Transfer
Name                                                               
Qiang Han       Peking University  Freshman   Male    87.0        N
Chengpeng Zhou   Fudan University    Senior   Male    81.0        N
Changpeng Zhao  Peking University  Freshman   Male    83.0        N
Chengpeng Qian   Fudan University    Senior   Male    73.0        Y

#### 练一练

select_dtypes 是一个实用函数，它能够从表中选出相应类型的列，

若要选出所有数值型的列，只需使用 .select_dtypes('number') ，

请利用布尔列表选择的方法

结合 DataFrame 的 dtypes 属性在 learn_pandas 数据集上实现这个功能。

In [70]:
df_test = df.copy()
df_test['add'] = range(0, df.shape[0])
# ---- 1 ------
df_test.select_dtypes('number')
# ----- 2 ------
condition1 = df_test.dtypes == 'float64'
condition2 = df_test.dtypes == 'int64'
df_test.loc[:, condition1 | condition2]

Weight  add
0      46.0    0
1      70.0    1
2      89.0    2
3      41.0    3
4      74.0    4
..      ...  ...
195    46.0  195
196    50.0  196
197    45.0  197
198    71.0  198
199    51.0  199

[200 rows x 2 columns]

In [74]:
# * 为函数，函数输入值为 df
def condition(x):
    condition_1_1 = x.School == 'Fudan University'
    condition_1_2 = x.Grade == 'Senior'
    condition_1_3 = x.Weight > 70
    condition_1 = condition_1_1 & condition_1_2 & condition_1_3
    condition_2_1 = x.School == 'Peking University'
    condition_2_2 = x.Grade == 'Senior'
    condition_2_3 = x.Weight > 80
    condition_2 = condition_2_1 & (~condition_2_2) & condition_2_3
    result = condition_1 | condition_2
    return result
print(df_demo.loc[condition])
print()
print(df_demo.loc[lambda x:'Quan Zhao', lambda x: 'Gender'])


                           School     Grade Gender  Weight Transfer
Name                                                               
Qiang Han       Peking University  Freshman   Male    87.0        N
Chengpeng Zhou   Fudan University    Senior   Male    81.0        N
Changpeng Zhao  Peking University  Freshman   Male    83.0        N
Chengpeng Qian   Fudan University    Senior   Male    73.0        Y

Female


### attention: 不要使用链式赋值

在对表或者序列赋值时，应当在使用一层索引器后直接进行赋值操作，

这样做是由于进行多次索引后赋值是赋在临时返回的 copy 副本上的，

而没有真正修改元素从而报出 SettingWithCopyWarning 警告。

例如，下面给出的例子：

In [80]:
df_chain = pd.DataFrame([[0,0],[1,0],[-1,0]], columns=list('AB'))
print(df_chain)
import warnings
with warnings.catch_warnings():
    warnings.filterwarnings('error')
    try:
        df_chain[df_chain.A != 0].B = 1 # 使用方括号列索引后，再使用点的列索引
    except Warning as w:
        Warnings_Msg = w
print(Warnings_Msg)

   A  B
0  0  0
1  1  0
2 -1  0

A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


# 3. iloc 索引器
> 针对位置筛选
>
> \* : 整数、整数列表、整数切片、布尔列表、函数
>


In [83]:
df_demo.iloc[1, 1]  # 第二行，第二列

'Freshman'

In [84]:
df_demo.iloc[[0, 1], [0, 1]]  # 前两行，前两列

School     Grade
Name                                                   
Gaopeng Yang    Shanghai Jiao Tong University  Freshman
Changqiang You              Peking University  Freshman

In [85]:
df_demo.iloc[1: 4, 2: 4]  # 不包含结束端点

Gender  Weight
Name                          
Changqiang You    Male    70.0
Mei Sun           Male    89.0
Xiaojuan Sun    Female    41.0

In [86]:
df_demo.iloc[lambda x: slice(1, 4)] 

School      Grade  Gender  Weight  \
Name                                                                       
Changqiang You              Peking University   Freshman    Male    70.0   
Mei Sun         Shanghai Jiao Tong University     Senior    Male    89.0   
Xiaojuan Sun                 Fudan University  Sophomore  Female    41.0   

               Transfer  
Name                     
Changqiang You        N  
Mei Sun               N  
Xiaojuan Sun          N

In [87]:
# 布尔筛选有限考虑 loc : iloc 不能传入 series 只能序列的values
df_demo.iloc[(df_demo.Weight > 80).values].head()

School      Grade Gender  Weight  \
Name                                                                      
Mei Sun         Shanghai Jiao Tong University     Senior   Male    89.0   
Qiang Zheng     Shanghai Jiao Tong University     Senior   Male    87.0   
Qiang Han                   Peking University   Freshman   Male    87.0   
Chengpeng Zhou               Fudan University     Senior   Male    81.0   
Feng Han        Shanghai Jiao Tong University  Sophomore   Male    82.0   

               Transfer  
Name                     
Mei Sun               N  
Qiang Zheng           N  
Qiang Han             N  
Chengpeng Zhou        N  
Feng Han              N

### 4. query  方法
> 传入： 字符串形式查询表达式
> 
> 执行结果：返回布尔列表
>
> 优点： 无需重复使用 DataFrame 来引用列名
>
> attention: 对于含有空格列名，使用 ‘col name’ 方式引用

In [91]:
df.query('((School == "Fudan University") &'
    ' (Grade == "Senior") &'
    ' (Weight > 70)) |'
    '((School == "Peking University") &'
    ' (Grade != "Senior") &'
    ' (Weight > 80))')

School     Grade            Name Gender  Weight Transfer
38   Peking University  Freshman       Qiang Han   Male    87.0        N
66    Fudan University    Senior  Chengpeng Zhou   Male    81.0        N
99   Peking University  Freshman  Changpeng Zhao   Male    83.0        N
131   Fudan University    Senior  Chengpeng Qian   Male    73.0        Y

In [93]:
# query ： 帮用户注册了来自 Dataframe 的列名，所有方法都可以被调用
df.query('Weight > Weight.mean()').head()

School      Grade            Name  Gender  Weight  \
1               Peking University   Freshman  Changqiang You    Male    70.0   
2   Shanghai Jiao Tong University     Senior         Mei Sun    Male    89.0   
4                Fudan University  Sophomore     Gaojuan You    Male    74.0   
10  Shanghai Jiao Tong University   Freshman   Xiaopeng Zhou    Male    74.0   
14            Tsinghua University     Senior    Xiaomei Zhou  Female    57.0   

   Transfer  
1         N  
2         N  
4         N  
10        N  
14        N

In [95]:
# 若干用法 ： or, and, is in, not in, between
df.query('(Grade not in["Freshman", "Sophomore"]) and'
         '(Gender == "Male")').head()
# == 等价于 is in

School   Grade           Name Gender  Weight  \
2   Shanghai Jiao Tong University  Senior        Mei Sun   Male    89.0   
16            Tsinghua University  Junior  Xiaoqiang Qin   Male    68.0   
17            Tsinghua University  Junior      Peng Wang   Male    65.0   
18            Tsinghua University  Senior   Xiaofeng Sun   Male    71.0   
21  Shanghai Jiao Tong University  Senior  Xiaopeng Shen   Male    62.0   

   Transfer  
2         N  
16        N  
17        N  
18        N  
21      NaN

In [97]:
df

School      Grade            Name  Gender  Weight  \
0    Shanghai Jiao Tong University   Freshman    Gaopeng Yang  Female    46.0   
1                Peking University   Freshman  Changqiang You    Male    70.0   
2    Shanghai Jiao Tong University     Senior         Mei Sun    Male    89.0   
3                 Fudan University  Sophomore    Xiaojuan Sun  Female    41.0   
4                 Fudan University  Sophomore     Gaojuan You    Male    74.0   
..                             ...        ...             ...     ...     ...   
195               Fudan University     Junior    Xiaojuan Sun  Female    46.0   
196            Tsinghua University     Senior         Li Zhao  Female    50.0   
197  Shanghai Jiao Tong University     Senior  Chengqiang Chu  Female    45.0   
198  Shanghai Jiao Tong University     Senior   Chengmei Shen    Male    71.0   
199            Tsinghua University  Sophomore     Chunpeng Lv    Male    51.0   

    Transfer  
0          N  
1          N  
2          N  
3          N  
4          N  
..       ...  
195        N  
196        N  
197        N  
198        N  
199        N  

[200 rows x 6 columns]

In [103]:
pd.__version__

'1.1.4'

In [105]:
# 若需要引入外部变量 变量名前加@
# low, high = 70, 80
# df.query('Weight.between(@low, @high)').head(5)

### 5. 随机抽样
> 对DataFrame：每一行看做一个样本，每一列看做一个特征，整个看做整体
>
> sample -- 对样本或特征进行随机抽样
>
> 统计特征来计算数据大致分布，耗费时间
>
> 多统计特征在等概率不放回的随机抽样条件下，是总体统计特征的无偏估计 
>
> 如样本均值，总体均值 -- 从表中抽取部分来做近似估计
>
> ==============================================
>
> sample: n, axis, frac, replace, weights 【抽样数量】【抽样方向】【抽样比例】【是否放回】【抽样相对概率】
>
>

In [111]:
# 以构造的value值相对大小作为抽样概率，做3次有放回抽样
df_sample = pd.DataFrame({'id': list('abcde'),
                          'value': [1, 2, 3, 4, 90]})
print(df_sample)
df_sample.sample(3, replace = True, weights = df_sample.value)

  id  value
0  a      1
1  b      2
2  c      3
3  d      4
4  e     90


id  value
4  e     90
4  e     90
4  e     90

# 多级索引
### 1. 多级索引及其表的结构

In [112]:
# 先构造一张表
np.random.seed(0)
multi_index = pd.MultiIndex.from_product([list('ABCD'),
                                         df.Gender.unique()], names = ('School', 'Gender'))
multi_column = pd.MultiIndex.from_product([['Height', 'Weight'],
                                          df.Grade.unique()], names = ('Indicator', 'Grade'))
df_multi = pd.DataFrame(np.c_[(np.random.rand(8, 4)*5 + 163).tolist(),
                              (np.random.randn(8, 4)*5 + 65).tolist()],
                       index = multi_index,
                       columns = multi_column).round(1)
df_multi
                                    

Indicator       Height                           Weight                   \
Grade         Freshman Senior Sophomore Junior Freshman Senior Sophomore   
School Gender                                                              
A      Female    165.7  166.6     166.0  165.7     76.3   57.7      65.2   
       Male      165.1  166.2     165.2  167.5     72.7   72.3      65.8   
B      Female    167.8  164.9     167.0  165.6     60.6   55.1      63.3   
       Male      165.8  167.6     163.4  163.4     71.2   71.0      63.1   
C      Female    163.1  167.2     166.9  167.4     59.8   57.9      56.5   
       Male      167.9  167.0     165.3  166.9     62.5   62.8      58.7   
D      Female    163.6  166.2     163.7  167.7     56.9   63.9      60.5   
       Male      165.6  165.1     164.3  166.9     62.4   59.1      64.9   

Indicator             
Grade         Junior  
School Gender         
A      Female   64.1  
       Male     66.9  
B      Female   65.8  
       Male     63.5  
C      Female   74.8  
       Male     68.9  
D      Female   66.9  
       Male     67.1

DataFrame 结构：
> 元素值， 行索引， 列索引 【索引：MultiIndex类型，其元素是元组】
>
> 行索引的第4个元素 ("B", "Male"), 列索引 2 ("Height", "Senior")a
>
> Attention: 外层连续出现相同的值时，第一次后出现的会被隐藏显示，增强可读性
>
>
>

![photo](https://upload-images.jianshu.io/upload_images/25252730-083bca7ed077da8f.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
>
>
> 图中的 School 和 Gender 分别对应了表的第一层和第二层行索引的名字， Indicator 和 Grade 分别对应了第一层和第二层列索引的名字
>
> 索引的名字和值属性都可以通过 names 和 values 获得


In [113]:
df_multi.index.names

FrozenList(['School', 'Gender'])

In [114]:
df_multi.columns.names

FrozenList(['Indicator', 'Grade'])

In [115]:
df_multi.index.values

array([('A', 'Female'), ('A', 'Male'), ('B', 'Female'), ('B', 'Male'),
       ('C', 'Female'), ('C', 'Male'), ('D', 'Female'), ('D', 'Male')],
      dtype=object)

In [116]:
df_multi.columns.values

array([('Height', 'Freshman'), ('Height', 'Senior'),
       ('Height', 'Sophomore'), ('Height', 'Junior'),
       ('Weight', 'Freshman'), ('Weight', 'Senior'),
       ('Weight', 'Sophomore'), ('Weight', 'Junior')], dtype=object)

若想要获得某一层索引

In [117]:
df_multi.index.get_level_values(0)

Index(['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D'], dtype='object', name='School')

对于索引，用户无法通过 index_obj\[0\] = item 来修改元素

### 2. 多级索引中的 loc 索引器

回到原表，将学校年级设为索引，行为多级索引，列为单级索引，默认状态的列索引不含名字，所以对应于刚刚图中Indicator和Grade的索引名位置空缺

In [118]:
df_multi = df.set_index(['School', 'Grade'])
df_multi.head()

Name  Gender  Weight  \
School                        Grade                                       
Shanghai Jiao Tong University Freshman     Gaopeng Yang  Female    46.0   
Peking University             Freshman   Changqiang You    Male    70.0   
Shanghai Jiao Tong University Senior            Mei Sun    Male    89.0   
Fudan University              Sophomore    Xiaojuan Sun  Female    41.0   
                              Sophomore     Gaojuan You    Male    74.0   

                                        Transfer  
School                        Grade               
Shanghai Jiao Tong University Freshman         N  
Peking University             Freshman         N  
Shanghai Jiao Tong University Senior           N  
Fudan University              Sophomore        N  
                              Sophomore        N

多级索引单个元素以元组为单位，照搬 loc 和 iloc, 建议对 MultiIndex进行排序避免性能警告

In [122]:
df_multi = df_multi.sort_index()
df_multi.loc[('Fudan University', 'Junior')].head()

Name  Gender  Weight Transfer
School           Grade                                         
Fudan University Junior      Yanli You  Female    48.0        N
                 Junior  Chunqiang Chu    Male    72.0        N
                 Junior   Changfeng Lv    Male    76.0        N
                 Junior     Yanjuan Lv  Female    49.0      NaN
                 Junior  Gaoqiang Zhou  Female    43.0        N

In [125]:
df_multi.loc[[('Fudan University', 'Senior'),
              ('Shanghai Jiao Tong University', 'Freshman')]]#链接两个表

Name  Gender  Weight  \
School                        Grade                                       
Fudan University              Senior    Chengpeng Zheng  Female    38.0   
                              Senior          Feng Zhou  Female    47.0   
                              Senior          Gaomei Lv  Female    34.0   
                              Senior          Chunli Lv  Female    56.0   
                              Senior     Chengpeng Zhou    Male    81.0   
                              Senior        Gaopeng Qin  Female    52.0   
                              Senior        Chunjuan Xu  Female    47.0   
                              Senior         Juan Zhang  Female    47.0   
                              Senior     Chengpeng Qian    Male    73.0   
                              Senior      Xiaojuan Qian  Female    50.0   
                              Senior            Quan Xu  Female    44.0   
Shanghai Jiao Tong University Freshman     Gaopeng Yang  Female    46.0   
                              Freshman        Qiang Chu  Female    52.0   
                              Freshman    Xiaopeng Zhou    Male    74.0   
                              Freshman       Yanpeng Lv    Male    65.0   
                              Freshman    Xiaopeng Zhao  Female    53.0   
                              Freshman      Chunli Zhao    Male    83.0   
                              Freshman       Peng Zhang  Female     NaN   
                              Freshman     Xiaoquan Sun  Female    40.0   
                              Freshman      Chunmei Shi  Female    52.0   
                              Freshman     Xiaomei Yang  Female    49.0   
                              Freshman    Xiaofeng Qian  Female    49.0   
                              Freshman      Changmei Lv    Male    75.0   
                              Freshman       Qiang Feng    Male    80.0   

                                       Transfer  
School                        Grade              
Fudan University              Senior          N  
                              Senior          N  
                              Senior          N  
                              Senior          N  
                              Senior          N  
                              Senior          N  
                              Senior          N  
                              Senior          N  
                              Senior          Y  
                              Senior          N  
                              Senior          N  
Shanghai Jiao Tong University Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N  
                              Freshman        N

In [126]:
df_multi.loc[df_multi.Weight > 70].head()  # 布尔列表

Name Gender  Weight Transfer
School           Grade                                           
Fudan University Freshman       Feng Wang   Male    74.0        N
                 Junior     Chunqiang Chu   Male    72.0        N
                 Junior      Changfeng Lv   Male    76.0        N
                 Senior    Chengpeng Zhou   Male    81.0        N
                 Senior    Chengpeng Qian   Male    73.0        Y

In [127]:
df_multi.loc[lambda x:('Fudan University', 'Junior')].head()

Name  Gender  Weight Transfer
School           Grade                                         
Fudan University Junior      Yanli You  Female    48.0        N
                 Junior  Chunqiang Chu    Male    72.0        N
                 Junior   Changfeng Lv    Male    76.0        N
                 Junior     Yanjuan Lv  Female    49.0      NaN
                 Junior  Gaoqiang Zhou  Female    43.0        N

练一练： 与单层索引列斯，若存在重复元素则不可以切片，请去除重复索引后给出元素切片的例子

In [135]:
df_multi.drop_duplicates()[1:].head()

Name  Gender  Weight Transfer
School           Grade                                          
Fudan University Freshman  Gaoqiang Qin  Female    63.0        N
                 Freshman  Gaofeng Zhao  Female    43.0        N
                 Freshman  Yanquan Wang  Female    55.0        N
                 Freshman     Feng Wang    Male    74.0        N
                 Freshman     Qiang Shi  Female    52.0        N

对多层的元素进行交叉组合后索引，同时指定 loc 的列全选则用 : 表示

形式为： `[(level_0_list, level_1_list), cols]`

In [136]:
# 想要所有北大和复旦的大二大三学生
res = df_multi.loc[(['Peking University', 'Fudan University'], 
                    ['Sophomore', 'Junior']), :]
print(res.shape)
res.head()

(33, 4)


Name  Gender  Weight Transfer
School            Grade                                           
Peking University Sophomore   Changmei Xu  Female    43.0        N
                  Sophomore  Xiaopeng Qin    Male     NaN        N
                  Sophomore        Mei Xu  Female    39.0        N
                  Sophomore   Xiaoli Zhou  Female    55.0        N
                  Sophomore      Peng Han  Female    34.0      NaN

In [139]:
# comparison
# 北大大三学生，复旦大二学生
res = df_multi.loc[[('Peking University', 'Junior'),
                    ('Fudan University', 'Sophomore')]]
print(res.shape)
res.head()

(16, 4)


Name  Gender  Weight Transfer
School            Grade                                         
Peking University Junior        Juan Xu  Female     NaN        N
                  Junior  Changjuan You  Female    47.0        N
                  Junior       Gaoli Xu  Female    48.0        N
                  Junior   Gaoquan Zhou    Male    70.0        N
                  Junior      Qiang You  Female    56.0        N

### 3. IndexSlice 对象



即使在索引不重复的时候，也能对元祖整体切片，而不能对每层切片，也不允许切片和布尔列表混合使用，引入的IndexSlice就是解决这个问题的。
下面构造一个索引不重复的DataFrame：
> `loc[idex[*, *]]`
>
> `loc[idx[*, *], idx[*, *]]`
>
>

> a. `loc[idx[*, *]]` 不能进行分层分别切片，前一个表示行，后一个是列
>
> b. `loc[idx[*, *], idx[*, *]]` 分层切片， idx前一个行，后一个列
>
>
>
>
>

In [142]:
np.random.seed(0)
L1, L2 = ['A', 'B', 'C'], ['a', 'b', 'c']
mul_index1 = pd.MultiIndex.from_product([L1, L2], names = ('Upper', 'Lower'))
L3, L4 = ['D', 'E', 'F'], ['d', 'e', 'f']
mul_index2 = pd.MultiIndex.from_product([L3, L4], names = ('Big', 'Small'))
df_ex = pd.DataFrame(np.random.randint(-9, 10, (9, 9)),
                    index = mul_index1, 
                    columns = mul_index2)
df_ex

Big          D        E        F      
Small        d  e  f  d  e  f  d  e  f
Upper Lower                           
A     a      3  6 -9 -6 -6 -2  0  9 -5
      b     -3  3 -8 -3 -2  5  8 -4  4
      c     -1  0  7 -4  6  6 -9  9 -6
B     a      8  5 -2 -9 -8  0 -9  1 -6
      b      2  9 -7 -9 -9 -5 -4 -3 -1
      c      8  6 -5  0  1 -8 -8 -2  0
C     a     -6 -3  2  5  9 -9  5 -6  3
      b      1  2 -5 -3 -5  6 -6  3 -5
      c     -1  5  6 -6  6  4  7  8 -4

In [145]:
idx = pd.IndexSlice
idx  # 为使用 slice 对象， 先定义

In [146]:
df_ex.loc[idx['C':, ('D', 'f'):]]

Big          D  E        F      
Small        f  d  e  f  d  e  f
Upper Lower                     
C     a      2  5  9 -9  5 -6  3
      b     -5 -3 -5  6 -6  3 -5
      c      6 -6  6  4  7  8 -4

In [147]:
# 布尔序列索引
df_ex.loc[idx[:'A', lambda x:x.sum() > 0]]

Big          D     F
Small        d  e  e
Upper Lower         
A     a      3  6  9
      b     -3  3 -4
      c     -1  0  9

In [148]:
df_ex.loc[idx[:'A', 'b':],idx['E':, 'e':]]

Big          E     F   
Small        e  f  e  f
Upper Lower            
A     b     -2  5 -4  4
      c      6  6  9 -6

### 多级索引的构造
with 多级索引表的结果和切片

如何构造多级索引
> set_index
>
>pd.MultiIndex
>
> --- 1. from_tuples
>
> --- 2. from_arrays
>
> --- 3. from_product
>
>
>

In [150]:
# from_tuples 根据传入由元组组成的列表进行构造
my_tuple = [('a', 'cat'), ('a', 'dog'), ('b', 'cat'), ('b', 'dog')]
pd.MultiIndex.from_tuples(my_tuple, names = ['First', 'Second'])

MultiIndex([('a', 'cat'),
            ('a', 'dog'),
            ('b', 'cat'),
            ('b', 'dog')],
           names=['First', 'Second'])

In [153]:
# from_arrays 根据传入列表中，对应层的列表进行构造
my_arrays = [list('aabb'), ['cat', 'dog'] * 2]
pd.MultiIndex.from_arrays(my_arrays, names = ['First', 'Second'])


MultiIndex([('a', 'cat'),
            ('a', 'dog'),
            ('b', 'cat'),
            ('b', 'dog')],
           names=['First', 'Second'])

In [154]:
# from_product 根据给定的多个列表的笛卡尔积进行构造
my_list1 = ['a', 'b']
my_list2 = ['cat', 'dog']
pd.MultiIndex.from_product([my_list1, 
                            my_list2],
                            names = ['First', 'Second'])


MultiIndex([('a', 'cat'),
            ('a', 'dog'),
            ('b', 'cat'),
            ('b', 'dog')],
           names=['First', 'Second'])

# 三、索引的常用方法

### 1. 索引层的交换和删除


In [156]:
# 构造一个三级索引的例子，方便理解交换过程
np.random.seed(0)
L1, L2, L3 = ['A', 'B'], ['a', 'b'], ['alpha', 'beta']
mul_index1 = pd.MultiIndex.from_product([L1, L2, L3], 
                                         names = ('Upper', 'Lower', 'Extra'))
L4, L5, L6 = ['C', 'D'], ['c', 'd'], ['cat', 'dog']
mul_index2 = pd.MultiIndex.from_product([L4, L5, L6], 
                                         names = ('Big', 'Small', 'Other'))
df_ex = pd.DataFrame(np.random.randint(-9, 10, (8, 8)),
                     index = mul_index1,
                     columns = mul_index2)
df_ex

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     alpha   3   6  -9  -6  -6  -2   0   9
            beta   -5  -3   3  -8  -3  -2   5   8
      b     alpha  -4   4  -1   0   7  -4   6   6
            beta   -9   9  -6   8   5  -2  -9  -8
B     a     alpha   0  -9   1  -6   2   9  -7  -9
            beta   -9  -5  -4  -3  -1   8   6  -5
      b     alpha   0   1  -8  -8  -2   0  -6  -3
            beta    2   5   9  -9   5  -6   3   1

索引层的交换由 swaplevel （只能交换两个层）和 reorder_levels（指定交换任意层） 完成（ 可以指定交换的是哪一个）

In [157]:
df_ex.swaplevel(0, 2, axis = 1).head()  # 列索引的第一层和第三层交换

Other             cat dog cat dog cat dog cat dog
Small               c   c   d   d   c   c   d   d
Big                 C   C   C   C   D   D   D   D
Upper Lower Extra                                
A     a     alpha   3   6  -9  -6  -6  -2   0   9
            beta   -5  -3   3  -8  -3  -2   5   8
      b     alpha  -4   4  -1   0   7  -4   6   6
            beta   -9   9  -6   8   5  -2  -9  -8
B     a     alpha   0  -9   1  -6   2   9  -7  -9

In [158]:
df_ex.reorder_levels([2, 0, 1], axis = 0).head() # 列表数组代表原来索引的层

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Extra Upper Lower                                
alpha A     a       3   6  -9  -6  -6  -2   0   9
beta  A     a      -5  -3   3  -8  -3  -2   5   8
alpha A     b      -4   4  -1   0   7  -4   6   6
beta  A     b      -9   9  -6   8   5  -2  -9  -8
alpha B     a       0  -9   1  -6   2   9  -7  -9

Attention： 轴之间的索引交换

行或列索引内部的交换，不同方向索引也可交换，后讨论


若要删除某一层的索引： droplevel 方法

In [160]:
df_ex.droplevel(1, axis = 1)

Big                 C               D            
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     alpha   3   6  -9  -6  -6  -2   0   9
            beta   -5  -3   3  -8  -3  -2   5   8
      b     alpha  -4   4  -1   0   7  -4   6   6
            beta   -9   9  -6   8   5  -2  -9  -8
B     a     alpha   0  -9   1  -6   2   9  -7  -9
            beta   -9  -5  -4  -3  -1   8   6  -5
      b     alpha   0   1  -8  -8  -2   0  -6  -3
            beta    2   5   9  -9   5  -6   3   1

In [161]:
df_ex.droplevel([0, 1], axis = 0)

Big     C               D            
Small   c       d       c       d    
Other cat dog cat dog cat dog cat dog
Extra                                
alpha   3   6  -9  -6  -6  -2   0   9
beta   -5  -3   3  -8  -3  -2   5   8
alpha  -4   4  -1   0   7  -4   6   6
beta   -9   9  -6   8   5  -2  -9  -8
alpha   0  -9   1  -6   2   9  -7  -9
beta   -9  -5  -4  -3  -1   8   6  -5
alpha   0   1  -8  -8  -2   0  -6  -3
beta    2   5   9  -9   5  -6   3   1

### 2. 索引属性的修改
> rename_axis 对索引层名字修改，修改方式是传入字典的映射
>
> 通过 rename 对索引值进行修改，如果是多集索引需要指定修改的层号
>

In [163]:
df_ex.rename_axis(index = {'Upper': 'Changed_row'}, 
                  columns = {'Other': 'Changed_Col'}).head()

Big                       C               D            
Small                     c       d       c       d    
Changed_Col             cat dog cat dog cat dog cat dog
Changed_row Lower Extra                                
A           a     alpha   3   6  -9  -6  -6  -2   0   9
                  beta   -5  -3   3  -8  -3  -2   5   8
            b     alpha  -4   4  -1   0   7  -4   6   6
                  beta   -9   9  -6   8   5  -2  -9  -8
B           a     alpha   0  -9   1  -6   2   9  -7  -9

In [164]:
df_ex.rename(columns = {'cat': 'not_cat'},
             level = 2).head()

Big                     C                       D                
Small                   c           d           c           d    
Other             not_cat dog not_cat dog not_cat dog not_cat dog
Upper Lower Extra                                                
A     a     alpha       3   6      -9  -6      -6  -2       0   9
            beta       -5  -3       3  -8      -3  -2       5   8
      b     alpha      -4   4      -1   0       7  -4       6   6
            beta       -9   9      -6   8       5  -2      -9  -8
B     a     alpha       0  -9       1  -6       2   9      -7  -9

In [165]:
# 传入参数也可以是函数，输入值就是索引元素
df_ex.rename(index = lambda x: str.upper(x),
             level = 2).head()

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     ALPHA   3   6  -9  -6  -6  -2   0   9
            BETA   -5  -3   3  -8  -3  -2   5   8
      b     ALPHA  -4   4  -1   0   7  -4   6   6
            BETA   -9   9  -6   8   5  -2  -9  -8
B     a     ALPHA   0  -9   1  -6   2   9  -7  -9

### 练一练

尝试在 rename_axis 中使用函数完成与例子中一样的功能。

In [175]:
# def replace(x):
#     return ['chage', 'kdf', 'dkflj']
# df_ex.rename_axis(index = replace)

对于整个索引的元素替换，利用迭代器实现


In [179]:
new_values = iter(list('abcdefgh'))  # 迭代对象
df_ex.rename(index = lambda x:next(new_values),
             level = 2)


Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     a       3   6  -9  -6  -6  -2   0   9
            b      -5  -3   3  -8  -3  -2   5   8
      b     c      -4   4  -1   0   7  -4   6   6
            d      -9   9  -6   8   5  -2  -9  -8
B     a     e       0  -9   1  -6   2   9  -7  -9
            f      -9  -5  -4  -3  -1   8   6  -5
      b     g       0   1  -8  -8  -2   0  -6  -3
            h       2   5   9  -9   5  -6   3   1

想要对某个位置元素进行修改，在单层索引时容易出现，先取出索引的 values 属性，再对得到的列表进行修改，最后对index对象重新赋值，。

对多级索引，一个解决方案是把某一层索引临时转为表的元素，然后在修改后，最后定回为索引

map 函数： 定义在index上的方法，传入的不是层的标量值，而是传入索引的元组，方便跨层修改


In [183]:
# 字符串转大写
df_tmp = df_ex.copy()
new_idx = df_tmp.index.map(lambda x:(x[0],
                                     x[1],
                                     str.upper(x[2])))
# 使用函数修改后，对应回到原来的位置
df_tmp.index = new_idx
df_tmp.head()

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     ALPHA   3   6  -9  -6  -6  -2   0   9
            BETA   -5  -3   3  -8  -3  -2   5   8
      b     ALPHA  -4   4  -1   0   7  -4   6   6
            BETA   -9   9  -6   8   5  -2  -9  -8
B     a     ALPHA   0  -9   1  -6   2   9  -7  -9

map 的另一个使用方法是多级索引的压缩


In [185]:
df_tmp = df_ex.copy()
new_idx = df_tmp.index.map(lambda x:(x[0] + '-' + 
                                     x[1] + '-' +
                                     x[2]))
# 实现了压缩
df_tmp.index = new_idx
df_tmp.head()  # 单层索引

Big         C               D            
Small       c       d       c       d    
Other     cat dog cat dog cat dog cat dog
A-a-alpha   3   6  -9  -6  -6  -2   0   9
A-a-beta   -5  -3   3  -8  -3  -2   5   8
A-b-alpha  -4   4  -1   0   7  -4   6   6
A-b-beta   -9   9  -6   8   5  -2  -9  -8
B-a-alpha   0  -9   1  -6   2   9  -7  -9

In [186]:
# 对压缩后的df，反向展开
new_idx =df_tmp.index.map(lambda x: tuple(x.split('-')))
df_tmp.index = new_idx
df_tmp.head()

Big         C               D            
Small       c       d       c       d    
Other     cat dog cat dog cat dog cat dog
A a alpha   3   6  -9  -6  -6  -2   0   9
    beta   -5  -3   3  -8  -3  -2   5   8
  b alpha  -4   4  -1   0   7  -4   6   6
    beta   -9   9  -6   8   5  -2  -9  -8
B a alpha   0  -9   1  -6   2   9  -7  -9

### 3. 索引的设置与重置


In [187]:
# 构造一个新表
df_new = pd.DataFrame({'A': list('aacd'),
                       'B': list('PQRT'),
                       'C': [1,2,3,4]})
df_new

A  B  C
0  a  P  1
1  a  Q  2
2  c  R  3
3  d  T  4

函数 set_index

参数 append：表示是否保留原来的索引，直接把新设定的添加到原索引内层


In [188]:
df_new.set_index('A')

B  C
A      
a  P  1
a  Q  2
c  R  3
d  T  4

In [189]:
df_new.set_index('A', append = True)

B  C
  A      
0 a  P  1
1 a  Q  2
2 c  R  3
3 d  T  4

In [190]:
# 同时指定多个列
df_new.set_index(['A', 'B'])

C
A B   
a P  1
  Q  2
c R  3
d T  4

In [197]:
# 若想要添加的索引不存在，可以直接传入对应 Series
my_index = pd.Series(list('WXYZ'), name = 'D')
df_new = df_new.set_index(['A', my_index])
df_new

B  C
A D      
a W  P  1
  X  Q  2
c Y  R  3
d Z  T  4

reset_index 是 set_index 的逆函数，主要参数是 drop，表示是否要把去掉的索引层丢弃，而不是添加



In [199]:
df_new.reset_index(['D'])


D  B  C
A         
a  W  P  1
a  X  Q  2
c  Y  R  3
d  Z  T  4

In [200]:
df_new.reset_index(['D'], drop = True) # 表示丢弃 'D'


B  C
A      
a  P  1
a  Q  2
c  R  3
d  T  4

In [201]:
# 重置所有索引，pandas生成默认索引
df_new.reset_index()

A  D  B  C
0  a  W  P  1
1  a  X  Q  2
2  c  Y  R  3
3  d  Z  T  4

### 4. 索引的变形

对索引做 扩充、剔除、添加新索引、旧索引对应元素填充到新索引的表


In [211]:
# 用旧表，做新表，添加一名新员工，去掉身高，添加性别
df_reindex = pd.DataFrame({"Weight":[60, 70, 80],
                           "Height":[176, 180, 179]},
                           index = ['1001', '1003', '1002'])
df_reindex

Weight  Height
1001      60     176
1003      70     180
1002      80     179

In [212]:
df_reindex.reindex(index = ['1001', '1003', '1003', '1004'],
                   columns = ['Weight', 'Gender'])


Weight  Gender
1001    60.0     NaN
1003    70.0     NaN
1003    70.0     NaN
1004     NaN     NaN

常用于时间序列索引的时间点填充，以及ID编号的扩充，原数据也会根据新表索引自动对其，与位置无关

In [213]:
# 相似函数 reindex_like,仿照传入表的索引，来进行被调用表索引的变形
df_existed = pd.DataFrame(index = ['1001', '1002', '1003', '1004'],
                          columns = ['Weight', 'Gender'])
df_reindex.reindex_like(df_existed)

Weight  Gender
1001    60.0     NaN
1002    80.0     NaN
1003    70.0     NaN
1004     NaN     NaN

## 五、练习  
### 余留：
**第一章练习批改未完成**

**第三章练习尚未独立完成，下次作业前一定完成**
### Ex1：公司员工数据集
现有一份公司员工数据集：

In [204]:
df = pd.read_csv('./data/company.csv')
df.head(3)

EmployeeID birthdate_key  age  city_name department      job_title gender
0        1318      1/3/1954   61  Vancouver  Executive            CEO      M
1        1319      1/3/1957   58  Vancouver  Executive      VP Stores      F
2        1320      1/2/1955   60  Vancouver  Executive  Legal Counsel      F

In [ ]:

1. 分别只使用`query`和`loc`选出年龄不超过四十岁且工作部门为`Dairy`或`Bakery`的男性。
2. 选出员工`ID`号 为奇数所在行的第1、第3和倒数第2列。
3. 按照以下步骤进行索引操作：

* 把后三列设为索引后交换内外两层
* 恢复中间一层
* 修改外层索引名为`Gender`
* 用下划线合并两层行索引
* 把行索引拆分为原状态
* 修改索引名为原表名称
* 恢复默认索引并将列保持为原表的相对位置

### Ex2：巧克力数据集
现有一份关于巧克力评价的数据集：

In [206]:
df = pd.read_csv('./data/chocolate.csv')
df.head(3)

Company  Review\nDate Cocoa\nPercent Company\nLocation  Rating
0  A. Morin          2016            63%            France    3.75
1  A. Morin          2015            70%            France    2.75
2  A. Morin          2015            70%            France    3.00

1. 把列索引名中的`\n`替换为空格。
2. 巧克力`Rating`评分为1至5，每0.25分一档，请选出2.75分及以下且可可含量`Cocoa Percent`高于中位数的样本。
3. 将`Review Date`和`Company Location`设为索引后，选出`Review Date`在2012年之后且`Company Location`不属于`France, Canada, Amsterdam, Belgium`的样本。